### PR TITLE
Add H5atclose and H5is_library_terminating routines.

### PR DESCRIPTION
**Pull request by qkoziol**
Adds two new API routines that provide useful infrastructure for third-party packages that wrap or extend HDF5.

### CODE REVIEW

**Comment by qkoziol** [comment]
RFC for these routines here: https://www.overleaf.com/read/jmmpfskybqdx

**Comment by qkoziol** [comment]
Figured out how to attach the RFC PDF:

[RFC_H5atclose_and_H5is_library_terminating.pdf](https://github.com/HDFGroup/hdf5/files/5611698/RFC_H5atclose_and_H5is_library_terminating.pdf)

**Comment by qkoziol** [comment]
Updated RFC with copy edits:

[RFC_H5atclose_and_H5is_library_terminating-2.pdf](https://github.com/HDFGroup/hdf5/files/5611709/RFC_H5atclose_and_H5is_library_terminating-2.pdf)

File: src/H5.c [hunk 1]
```diff
@@ -1124,6 +1198,34 @@ H5is_library_threadsafe(hbool_t *is_ts)
...
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5is_library_terminating(hbool_t *is_terminating)
+{
+    FUNC_ENTER_API_NOINIT
+    H5TRACE1("e", "*b", is_terminating);
+
+    HDassert(is_terminating);
```
Comment (derobins) [inline src/H5.c:1226]: Should use real error checking and not assert() in public API calls.

File: src/H5public.h [hunk 2]
```diff
@@ -386,6 +386,9 @@ typedef struct H5_alloc_stats_t {
     size_t             peak_alloc_blocks_count;  /**< Peak # of blocks allocated */
 } H5_alloc_stats_t;
 
+/* Library shutdown callback */
+typedef void (*H5_atclose_func_t)(void *ctx);
```
Comment (derobins) [inline src/H5public.h:392]: Do typedefs get doxygen comments?

File: src/H5.c [hunk 3]
```diff
@@ -88,6 +95,12 @@ H5_debug_t     H5_debug_g; /* debugging info */
 /* Local Variables */
 /*******************/
 
+/* Linked list of registered 'atclose' functions to invoke at library shutdown */
+static H5_atclose_node_t *H5_atclose_head = NULL;
+
+/* Declare a free list to manage the H5_atclose_node_t struct */
+H5FL_DEFINE_STATIC(H5_atclose_node_t);
```
Comment (derobins) [inline src/H5.c:102]: I feel like using a free list here is overkill. I seriously doubt there's going to be a lot of memory churn on close callbacks.

**Comment by derobins** [comment]
I would approve this, but I feel like the assert-based error check needs to be fixed first.

File: src/H5.c (hunk 1 above)
Comment (qkoziol) [reply src/H5.c:1226]: That's not possible here, since the library may not be initialized.

File: src/H5public.h (hunk 2 above)
Comment (qkoziol) [reply src/H5public.h:392]: Seems like it (based on H5Fpublic.h).   Added.

**Comment by qkoziol** [comment]
Updated RFC with formatting and funding acknowledgement:

[RFC_H5atclose_and_H5is_library_terminating-3.pdf](https://github.com/HDFGroup/hdf5/files/5612364/RFC_H5atclose_and_H5is_library_terminating-3.pdf)

File: src/H5.c (hunk 1 above)
Comment (derobins) [reply src/H5.c:1226]: But you can still check the parameter and jump to the end returning FAIL if the parameter is NULL.

Comment (qkoziol) [reply src/H5.c:1226]: OK, sure - I've changed the code to do that for both H5is_library_terminating as well as the existing H5is_library_threadsafe routine.

**Review by derobins — APPROVED** [review]

**Comment by qkoziol** [comment]
This PR needs one more review, from either: @soumagne @fortnern @jhendersonHDF @vchoi-hdfgroup or @jrmainzer    Please take a look, so I can move forward with other PRs.

**Comment by qkoziol** [comment]
This is the last of my outstanding PRs, can I please get another review, so I can clear my plate and start on the next set?

File: src/H5.c (hunk 3 above)
Comment (soumagne) [reply src/H5.c:102]: also tend to agree with Dana

**Review by soumagne — APPROVED** [review]

**Pull request merged**